File: src/ReadStorm.Infrastructure/Services/RuleBasedSourceDiagnosticUseCase.cs
using AngleSharp.Html.Parser;
using ReadStorm.Application.Abstractions;
using ReadStorm.Domain.Models;

namespace ReadStorm.Infrastructure.Services;

public sealed class RuleBasedSourceDiagnosticUseCase : ISourceDiagnosticUseCase
{
    private readonly IReadOnlyList<string> _ruleDirectories;
    private readonly HttpClient _httpClient;

    public RuleBasedSourceDiagnosticUseCase(
        HttpClient? httpClient = null,
        IReadOnlyList<string>? ruleDirectories = null)
    {
        _ruleDirectories = ruleDirectories ?? RulePathResolver.ResolveAllRuleDirectories();
        _httpClient = httpClient ?? RuleHttpHelper.CreateHttpClient(TimeSpan.FromSeconds(5));
    }

    public async Task<SourceDiagnosticResult> DiagnoseAsync(
        int sourceId,
        string testKeyword,
        CancellationToken cancellationToken = default)
    {
        // 每个书源诊断最多 8 秒
        using var timeoutCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
        timeoutCts.CancelAfter(TimeSpan.FromSeconds(8));
        var ct = timeoutCts.Token;

        var result = new SourceDiagnosticResult { SourceId = sourceId };

        void Log(string message) => result.DiagnosticLines.Add($"[{DateTimeOffset.Now:HH:mm:ss.fff}] {message}");

        try
        {
            Log($"开始诊断书源 {sourceId}");

            var rule = await LoadRuleAsync(sourceId, ct);
            if (rule is null)
            {
                result.Summary = $"未找到书源 {sourceId} 的规则文件。";
                Log(result.Summary);
                return result;
            }

            result.SourceName = rule.Name ?? $"Rule-{rule.Id}";
            result.BaseUrl = rule.Url ?? string.Empty;

            // Check search rule
            result.SearchRuleFound = rule.Search is not null
                                     && !string.IsNullOrWhiteSpace(rule.Search.Url);
            Log($"搜索规则：{(result.SearchRuleFound ? "已配置" : "缺失")}");

            // Check toc rule
            result.TocRuleFound = rule.Toc is not null
                                  && !string.IsNullOrWhiteSpace(rule.Toc.Item);
            result.TocSelector = rule.Toc?.Item ?? string.Empty;
            Log($"目录规则：{(result.TocRuleFound ? "已配置" : "缺失")}，selector='{result.TocSelector}'");

            // Check chapter rule
            result.ChapterRuleFound = rule.Chapter is not null
                                      && !string.IsNullOrWhiteSpace(rule.Chapter.Content);
            result.ChapterContentSelector = rule.Chapter?.Content ?? string.Empty;
            Log($"章节规则：{(result.ChapterRuleFound ? "已配置" : "缺失")}，selector='{result.ChapterContentSelector}'");

            // HTTP connectivity check
            if (!string.IsNullOrWhiteSpace(result.BaseUrl))
            {
                try
                {
                    using var request = new HttpRequestMessage(HttpMethod.Head, result.BaseUrl);
                    using var response = await _httpClient.SendAsync(request, ct);
                    result.HttpStatusCode = (int)response.StatusCode;
                    result.HttpStatusMessage = response.StatusCode.ToString();
                    Log($"HTTP 连通性：{result.HttpStatusCode} {result.HttpStatusMessage}");
                }
                catch (Exception ex)
                {
                    result.HttpStatusCode = 0;
                    result.HttpStatusMessage = ex.GetType().Name + ": " + ex.Message;
                    Log($"HTTP 连通性检测失败：{result.HttpStatusMessage}");
                }
            }

            // Search test
            if (result.SearchRuleFound && !string.IsNullOrWhiteSpace(testKeyword))
            {
                try
                {
                    var searchUrl = rule.Search!.Url!.Replace("%s", Uri.EscapeDataString(testKeyword));
                    Log($"尝试搜索：{searchUrl}");
                    using var searchRequest = new HttpRequestMessage(HttpMethod.Get, searchUrl);
                    using var searchResponse = await _httpClient.SendAsync(searchRequest, ct);
                    if (searchResponse.IsSuccessStatusCode)
                    {
                        var html = await searchResponse.Content.ReadAsStringAsync(ct);
                        var resultSelector = RuleFileLoader.NormalizeSelector(rule.Search.Result);
                        if (!string.IsNullOrWhiteSpace(resultSelector))
                        {
                            var parser = new HtmlParser();
                            var doc = parser.ParseDocument(html);
                            result.SearchResultCount = doc.QuerySelectorAll(resultSelector).Length;
                            Log($"搜索结果 selector '{resultSelector}' 命中：{result.SearchResultCount} 条");
                        }
                        else
                        {
                            Log("搜索结果 selector 为空，无法解析。");
                        }
                    }
                    else
                    {
                        Log($"搜索请求返回：{(int)searchResponse.StatusCode}");
                    }
                }
                catch (Exception ex)
                {
                    Log($"搜索测试异常：{ex.Message}");
                }
            }

            result.Summary = result.IsHealthy ? "书源状态正常" : "书源存在配置或连通问题";
            Log($"诊断结论：{result.Summary}");
        }
        catch (Exception ex)
        {
            result.Summary = $"诊断异常：{ex.Message}";
            Log(result.Summary);
        }

        return result;
    }

    private Task<RuleFileDto?> LoadRuleAsync(int sourceId, CancellationToken cancellationToken)
        => RuleFileLoader.LoadRuleAsync(_ruleDirectories, sourceId, cancellationToken);
}

File: pureopus/app/src/main/java/com/readstorm/app/infrastructure/services/RuleBasedSourceDiagnosticUseCase.kt
package com.readstorm.app.infrastructure.services

import android.content.Context
import com.readstorm.app.application.abstractions.ISourceDiagnosticUseCase
import com.readstorm.app.domain.models.SourceDiagnosticResult
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request

class RuleBasedSourceDiagnosticUseCase(
    private val context: Context
) : ISourceDiagnosticUseCase {

    private val httpClient: OkHttpClient = RuleHttpHelper.createHttpClient(timeoutSeconds = 8)

    override suspend fun diagnose(sourceId: Int, testKeyword: String): SourceDiagnosticResult =
        withContext(Dispatchers.IO) {
            val result = SourceDiagnosticResult(sourceId = sourceId)

            try {
                val rule = RuleFileLoader.loadRule(context, sourceId)
                if (rule == null) {
                    result.diagnosticLines.add("❌ 未找到 sourceId=$sourceId 的规则文件")
                    return@withContext result
                }

                result.sourceName = rule.name
                result.baseUrl = rule.url

                // Check rule sections
                result.searchRuleFound = rule.search != null
                result.tocRuleFound = rule.toc != null
                result.chapterRuleFound = rule.chapter != null

                result.diagnosticLines.add("📋 书源: ${rule.name} (id=$sourceId)")
                result.diagnosticLines.add("🔗 基础URL: ${rule.url}")
                result.diagnosticLines.add("🔎 搜索规则: ${if (result.searchRuleFound) "✅" else "❌"}")
                result.diagnosticLines.add("📑 目录规则: ${if (result.tocRuleFound) "✅" else "❌"}")
                result.diagnosticLines.add("📖 章节规则: ${if (result.chapterRuleFound) "✅" else "❌"}")

                // Test HTTP connectivity
                if (rule.url.isNotBlank()) {
                    try {
                        val request = Request.Builder().url(rule.url).build()
                        val response = httpClient.newCall(request).execute()
                        response.use { resp ->
                            result.httpStatusCode = resp.code
                            result.httpStatusMessage = resp.message
                            result.diagnosticLines.add("🌐 HTTP状态: ${resp.code} ${resp.message}")
                        }
                    } catch (e: Exception) {
                        result.diagnosticLines.add("🌐 HTTP连接失败: ${e.message}")
                    }
                }

                // Test search if available
                if (rule.search != null && testKeyword.isNotBlank()) {
                    try {
                        val searchUseCase = RuleBasedSearchBooksUseCase(context)
                        val results = searchUseCase.execute(testKeyword, sourceId)
                        result.searchResultCount = results.size
                        result.diagnosticLines.add("🔍 搜索结果: ${results.size} 条")
                    } catch (e: Exception) {
                        result.diagnosticLines.add("🔍 搜索失败: ${e.message}")
                    }
                }

                // Report TOC selector
                if (rule.toc != null && rule.toc!!.item.isNotBlank()) {
                    result.tocSelector = rule.toc!!.item
                    result.diagnosticLines.add("📑 目录选择器: ${rule.toc!!.item}")
                }

                if (rule.chapter != null && rule.chapter!!.content.isNotBlank()) {
                    result.chapterContentSelector = rule.chapter!!.content
                    result.diagnosticLines.add("📖 内容选择器: ${rule.chapter!!.content}")
                }

            } catch (e: Exception) {
                result.diagnosticLines.add("❌ 诊断异常: ${e.message}")
            }

            result
        }
}
